core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -85,21 +88,24 @@
     private final DiscoverySettings discoverySettings;
     private final ClusterName clusterName;
     private final PendingClusterStatesQueue pendingStatesQueue;
+    private final CustomPrototypeRegistry registry;
 
     public PublishClusterStateAction(
             Settings settings,
             TransportService transportService,
             Supplier<ClusterState> clusterStateSupplier,
             NewPendingClusterStateListener listener,
             DiscoverySettings discoverySettings,
-            ClusterName clusterName) {
+            ClusterName clusterName,
+            CustomPrototypeRegistry registry) {
```

Comment:
@bleskes Like we discussed last friday I tried to stop using named writeables for cluster state related stuff and use the registry directly (so that all code will follow the same deserialization logic), but I ran into an issue that I don't know how to solve without creating a big mess. There are a couple of places where we serialize cluster related stuff in the action framework:
* `ClusterRerouteResponse#readFrom(...)`
* `ClusterStateResponse#readFrom(...)`
* `CreateIndexRequest#readFrom(...)`
* `PutIndexTemplateRequest#readFrom(...)`

In these readFrom() methods I don't have access to the registry. I don't see a way of passing down the registry without changing the `Streamable` interface.